API changes to support profile privacy options

ahukkanen opened the pull request:
#### :tophat: What? Why?
In order to support profile privacy in the external module we are building right now, we will add any proposed API changes to this PR.

This is the second iteration after the following findings:
1. It would be beneficial also for the core to consider the user profile publicity based on if the user has accepted the terms of service which might be a requirement for displaying their profile on the website when they first sign in to the platform.
2. It seems a better approach to consider the profile publicity only in places where we would actually display the user data publicly on the website. This approach seems to require changes in less modules (including 3rd party modules) making it more convenient overall.

~~This is still work in progress but I will open this already now so that we can continue on the same PR.~~

This is now ready for review.

#### :pushpin: Related Issues
- Supersedes #10939
- Related to
  * https://meta.decidim.org/processes/roadmap/f/122/proposals/16247
  * https://meta.decidim.org/processes/roadmap/f/122/proposals/17297
  * https://github.com/mainio/decidim-module-privacy

#### Testing

- Register to the platform as a new user
- See that straight after the registration, you cannot access any of the user's public profile pages (should give a 404)
- See that the user is not visible in the API
- See that managed users do not appear in the API
- Confirm the newly created user profile using the link in the email
- See that you can now access the profile page and see the user details through the API
- Manually change the user in the database to have not accepted the terms of service
- See that the public profile pages are again hidden and the user is not visible through the API

To test what happens in case all user profiles are not published:

- Make a local change to the code: change the [`profile_published?`](https://github.com/decidim/decidim/blob/aa3eb43f655f7fc36830cf193dbce83b8ee1dc92/decidim-core/app/models/decidim/user.rb#L92) method to return `false` always
- See that you can no longer see the details of these users on the website
- See that the "public profile" link is now gone under the user's personal links
- See that the user details are hidden from the API
  * Note that you will still get a list of users through the API using this method because the API query still returns the whole list of confirmed users but the details of these users should be hidden even in this case

ahukkanen committed:
Implement profile publicity scopes

ahukkanen committed:
Consider visibility in the author display and in data sources

- Fix issues displaying the author cell when the profile is not
  visible
- Hide the author from the API if the profile is not visible
- Hide the author from the data exports if the profile is not
  visible

ahukkanen committed:
Move the nickname and email validations to the user base entity

ahukkanen committed:
Do not display assembly member name when they are not visible

ahukkanen committed:
Display the user group name for a member of the group

ahukkanen committed:
Allow user group members to access the profile page

ahukkanen committed:
Fix disambiguous column name

ahukkanen committed:
Fix the user base entity validations

ahukkanen committed:
Prevent N+1 queries when checking for tos acceptance

ahukkanen committed:
Fix seeds

ahukkanen committed:
Fix specs and factories

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Fix managed users error link to managed user profile path

ahukkanen committed:
Fix author cell for deleted users

ahukkanen committed:
Fix email and nickname validations

ahukkanen committed:
Fix the visible memberships references for groups

ahukkanen committed:
Lint

ahukkanen committed:
Fix the official author presenter

ahukkanen committed:
Fix hidden user profile path/url and show generated name for deleted users

ahukkanen committed:
Add the tos_agreement back to the seeds

ahukkanen committed:
Return the original author record if the author is deleted

ahukkanen committed:
Fix more specs

ahukkanen committed:
Fix seeds for blogs component

ahukkanen committed:
Fix more specs

ahukkanen committed:
Fix more specs

ahukkanen committed:
Fix more core specs

ahukkanen committed:
Fix the create assembly member spec

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

alecslupu commented:
@ahukkanen maybe you could merge this with the decidim develop as well? I see there are many files that seems to be added, but they are already in develop.

alecslupu commented:
> @ahukkanen maybe you could merge this with the decidim develop as well? I see there are many files that seems to be added, but they are already in develop.

nevermind! You already updated it

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Fix comments specs

ahukkanen committed:
Fix explore blog posts index specs

andreslucena commented:
@ahukkanen I'm checking out old PRs and found this one. Are you planning on retaking this?

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Call the `.presenter` method only when author is returned

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

andreslucena commented:
Hi @ahukkanen 

I'm trying to move some old PRs, can we retake this PR? 
 
> @andreslucena Definitely this is very important for us and I will finish this as soon as we can get feedback on the remaining issue, i.e. discussion #11057.
> 
> There are few failing specs because of this open issue and I don't know how to handle that before we get feedback from @decidim/product on what kind of approach to take with the groups visibility.
> 
> Other than that, this is already pretty much done.

About the doubt of the groups, after the last changes mentioned by @carolromero on the comment https://github.com/decidim/decidim/discussions/11057#discussioncomment-13192443 

> It seems a better approach to consider the profile publicity only in places where we would actually display the user data publicly on the website. This approach seems to require changes in less modules (including 3rd party modules) making it more convenient overall.

Slightly related to this one, we're starting the work on minimizing the data on things like the nickname (and hence, the visibility of the profile) when the participant is doing actions like only voting and not having a reason to actually have a nickname (like in cases where a participant creates a proposal, meeting, comment, etc). Check out https://meta.decidim.org/processes/roadmap/f/122/proposals/17898

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Make users `:confirmed` in specs that have been renamed

These were not recognized by the git merge.

ahukkanen committed:
Re-add the nickname validation (removed in the merge commit)

ahukkanen committed:
Remove references to `Decidim::UserGroup`

ahukkanen committed:
Fix remaining merge conflicts

ahukkanen committed:
Re-add the methods to the `User` model that were removed in the merge

ahukkanen committed:
Lint

ahukkanen committed:
Re-add a `User` model scope removed during the merge

ahukkanen committed:
Fix broken specs

ahukkanen committed:
Fix checking if the author is deleted (does not work for organization)

ahukkanen committed:
Return the author as-is in case it is an organization

ahukkanen committed:
Do not return a dummy author if the author does not exist

ahukkanen committed:
Fix the cancel coauthorship spec

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Fix conference speaker presenter spec

ahukkanen commented:
@andreslucena I believe this is now ready for review and conflicts are resolved.

I believe the failing pipelines are not related to this PR since the two failing ones passed on the previous commit and the last commit did not change anything related to those specs (only conferences).

> About the doubt of the groups, after the last changes mentioned by @carolromero on the comment https://github.com/decidim/decidim/discussions/11057#discussioncomment-13192443

The question about groups is no longer relevant since the removal of groups (#14130).

ahukkanen committed:
Do not display the user activities page if the profile is not public

ahukkanen committed:
Fix the organization on the dummy author user

ahukkanen committed:
Only require `Decidim::User` type of users to be confirmed within the scope

ahukkanen committed:
Do not display managed users within the API

ahukkanen committed:
Fetch the user name within the API through the presenter

In order to hide the name in case the user is not public.

ahukkanen committed:
Do not display a link to the public profile if the profile is not public

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

andreslucena commented:
Passing the review to @alecslupu as it's a big review and I will not have time in the next weeks for this.

alecslupu commented on decidim-conferences/spec/presenters/decidim/admin/conference_speaker_presenter_spec.rb:
```diff
@@ -4,8 +4,9 @@
 
 module Decidim
   describe Admin::ConferenceSpeakerPresenter, type: :helper do
+    let(:conference) { create(:conference, skip_injection: true) }
```
Why are you skipping the injection?

ahukkanen replied:
Unsure.

ahukkanen replied:
9bd896e4015b973753a776ee58d9e42c9bc9db9f

alecslupu commented on decidim-conferences/spec/presenters/decidim/conference_speaker_presenter_spec.rb:
```diff
@@ -4,8 +4,9 @@
 
 module Decidim
   describe ConferenceSpeakerPresenter, type: :helper do
+    let(:conference) { create(:conference, skip_injection: true) }
```
Why are you skipping the injection?

ahukkanen replied:
Unsure.

ahukkanen replied:
9bd896e4015b973753a776ee58d9e42c9bc9db9f

alecslupu commented on decidim-proposals/lib/decidim/proposals/seeds.rb:
I think we have the same thing into
- `decidim-meetings/lib/decidim/meetings/seeds.rb:133`
- `decidim-debates/lib/decidim/debates/seeds.rb:15`

ahukkanen replied:
I can add it to meetings seeds. Unsure where this needs to be added at the debates seeds.

andreslucena replied:
@alecslupu you should do links with permalinks when referring to lines in another files, as you can be sure that the lines in a live codebase will change with time for sure.

I guess he was referring to 

https://github.com/decidim/decidim/blob/bcf09dc08a5ff48db5d059d358427db5cbfe3374/decidim-debates/lib/decidim/debates/seeds.rb#L15-L18

But as it is doing find for user@example.org I understand that this isn't actually needed

alecslupu commented on decidim-core/spec/controllers/omniauth_registrations_controller_spec.rb:
```diff
@@ -34,13 +34,9 @@ module Decidim
         context "when the given resource is a user" do
           context "and is an admin" do
-            let(:user) { build(:user, :admin, sign_in_count: 1) }
+            let(:user) { build(:user, :admin, sign_in_count: 1, organization:) }
```
Maybe we need to add `:confirmed` ?

ahukkanen replied:
Done.

alecslupu commented on decidim-core/app/models/decidim/user_base_entity.rb:
```diff
@@ -32,12 +32,50 @@ class UserBaseEntity < ApplicationRecord
+    # User entities (user groups) that have their profile visible on the
+    # platform from the user's own perspective. This includes also blocked users
+    # because the decision to hide their profile has been made by someone else.
+    scope :profile_published, -> { confirmed.where(deleted_at: nil, managed: false) }
```
Maybe would better to apply scopes here, while we move managed / not_managed scopes to parent class? like 
```suggestion
    scope :profile_published, -> { confirmed.not_deleted.not_managed }
```

ahukkanen replied:
`not_managed` is defined at `Decidim::User`. The `Decidim::UserBaseEntity` does not define that scope.

I will implement the other change.

alecslupu commented on decidim-core/app/models/decidim/user.rb:
```diff
@@ -77,8 +79,8 @@ def self.all
-                      index_on_create: ->(user) { !(user.deleted? || user.blocked?) },
-                      index_on_update: ->(user) { !(user.deleted? || user.blocked?) })
+                      index_on_create: ->(user) { !user.hidden? },
+                      index_on_update: ->(user) { !user.hidden? })
```
why not
```suggestion
                      index_on_create: ->(user) { user.visible? },
                      index_on_update: ->(user) { user.visible? })
```

alecslupu commented on decidim-core/app/models/decidim/user.rb:
```diff
@@ -49,6 +49,8 @@ def self.all
 
     has_one_attached :download_your_data_file
 
+    scope :tos_accepted, -> { where.not(accepted_tos_version: nil) }
```
I think this may be valid for all types of users?

ahukkanen replied:
At least in the past wasn't since group users did not have to accept TOS.

Unsure what is the current situation.

andreslucena replied:
I think its safe to have this, as we could still have Users that came from UserGroups with ToS not accepted. 

I checked in Metadecidim and https://meta.decidim.org/en/profiles/product/activity has an accepted_tos_version of "2025-10-30 10:51:20.539616000 +0000" - that matches the v0.31.0.rc2 update at Metadecidim, so I understand that what happened is that the NeedsTosAccepted kicked in on first login.

alecslupu commented on decidim-core/app/cells/decidim/user_profile_cell.rb:
```diff
@@ -20,6 +20,8 @@ def role
     end
 
     def resource_path
+      return "#" unless user.visible?
```
```suggestion
      return "#" if user.hidden?
```

alecslupu reviewed (changes requested):
I have reviewed 148 / 166 . Can you sink with the latest develop?

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Apply suggestions from code review

Co-authored-by: Alexandru Emil Lupu <contact@alecslupu.ro>

ahukkanen committed:
Address misc feedback

andreslucena commented:
@alecslupu have you seen Antti'ß comments? Can we try to retake this one too? Thanks

alecslupu commented:
@andreslucena , as you can see the pipeline is quite red, 

I am waiting for @ahukkanen to address the issues with the pipeline.

<img width="970" height="504" alt="image" src="https://github.com/user-attachments/assets/c14073f4-cacf-4b28-851f-8fa39aec96ee" />

alecslupu commented:
@ahukkanen can you sync with latest develop?

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Fix presenters for user, member and speaker

ahukkanen committed:
Make ApiUser profile published always

ahukkanen committed:
Fix welcome notification if profile is not published

ahukkanen committed:
Fix expectation in participatory space members examples

ahukkanen committed:
Fix user presenter spec and add test for non-confirmed user name

ahukkanen committed:
Fix author cell spec

ahukkanen committed:
Remove method no longer used

ahukkanen committed:
Add missing helper to ConferenceSpeakerPresenter

ahukkanen committed:
Return mapped author if the author is a group

ahukkanen committed:
Fix more specs not considering the visibility checks before

ahukkanen committed:
Fix proposals data migration spec (user confirmed)

ahukkanen committed:
Apply review recommendations to conference speaker specs

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Fix specs

ahukkanen committed:
Fix more specs

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

andreslucena commented:
@ahukkanen we're trying to lower the number of open PRs and this one is really old. I see that the CI is passing. Can we do anything to move this forward? Thanks!

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen commented:
> @ahukkanen we're trying to lower the number of open PRs and this one is really old. I see that the CI is passing. Can we do anything to move this forward? Thanks!

To review it?

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Fix spec for ContentRenderers::ResourceRenderer

andreslucena commented on decidim-core/lib/decidim/authorable.rb:
```diff
@@ -38,6 +38,20 @@ module Authorable
+        return mapped_author if mapped_author.try(:deleted?)
+        return mapped_author if mapped_author.try(:ephemeral?)
+
+        Decidim::User.new(id: 0, name: "", nickname: "", decidim_organization_id: mapped_author.decidim_organization_id)
```
Just to open the discussion as I'm not sure if there are other places where this would make sense, but just to be consistent should we introduce the same pattern as `Decidim::NilPresenter` here? I'm referring to having its own class with the methods that should answer.

ahukkanen replied:
I would keep it consistent as I don't know what consequences there would be to return another object than user or organization here (i.e. an ActiveRecord model). Returning the organization itself would not make sense here since it would change how the author is presented possibly causing confusion.

If I remember correctly, the original reason for adding this dummy return object was to get around the `author_belongs_to_organization` validation. But that would not necessarily be the only place that could break as e.g. here it is assuming this method to return a record responding to `#name`:
https://github.com/decidim/decidim/blob/50261af4d7d9a4ac47e402baff15544e5fb4c3d3/decidim-blogs/lib/decidim/blogs/schema_org_blog_posting_post_serializer.rb#L49

That could be worked around by implementing that method in the presenter but changing the method to return a presenter would change its signature, i.e. the return object type would change from `ActiveRecord::Base`.

andreslucena replied:
Ok, that makes sense!

andreslucena commented on decidim-core/lib/decidim/api/types/user_type.rb:
```diff
@@ -63,7 +67,7 @@ def badges
 
       def self.authorized?(object, context)
-        super && object.confirmed? && !object.blocked? && !object.deleted?
+        super && object.visible?
```
Related to this change, I found other places where we should replace these permissions checks and introduce the new pattern of `visible`: 

```
rg confirmed |rg blocked| rg deleted
decidim-comments/app/models/decidim/comments/seed.rb:          user = Decidim::User.where(organization:).not_deleted.not_blocked.confirmed.sample
decidim-core/app/queries/decidim/stats_users_count.rb:      users = Decidim::User.where(organization: @organization).not_deleted.not_blocked.confirmed
decidim-core/app/queries/decidim/authorized_users.rb:      @users = organization.users.not_deleted.not_blocked.confirmed
decidim-core/lib/tasks/upgrade/user_groups_migration.rake:            next if user.deleted? || user.blocked? || !user.confirmed?
decidim-core/lib/tasks/upgrade/user_groups_migration.rake:            next if user.deleted? || user.blocked? || !user.confirmed?
```

(The data migrations upgrade tasks can be ignored, as they should have been already ran a couple upgrades ago)

ahukkanen replied:
I would be a more careful about these suggested changes as `User.visible` also excludes managed users on top of these conditions. Changing these would therefore change behavior in all of the suggested places.

andreslucena replied:
Oh nice catch, I agree!

andreslucena commented on decidim-core/app/models/decidim/user_base_entity.rb:
```diff
@@ -32,12 +32,50 @@ class UserBaseEntity < ApplicationRecord
+      )
+    }
+
+    # User entities (user groups) that have their profile visible on the
```
We don't have user groups anymore so I'd just remove mentioning them if not necessary (as it just noise to newcomers as far as  I understand)

ahukkanen replied:
7cab1b2a4ccd8ddcf1ba6e7440fb701cd0be267c

andreslucena commented on decidim-core/app/presenters/decidim/personal_user_presenter.rb:
```diff
@@ -0,0 +1,21 @@
+  # Decorator for users when they can see the user details themselves, i.e.
+  # when the user is logged in themselves.
+  #
+  class PersonalUserPresenter < UserPresenter
```
I don't like the Personal wording, as it is a new concept. 

I see that this is being used at decidim-core/lib/decidim/api/types/session_type.rb, so maybe `SessionUserPresenter` or `CurrentUserPresenter` ?

ahukkanen replied:
477a0649b06357121aaa94f58f861fd85d91c2e3

andreslucena commented on decidim-proposals/spec/permissions/decidim/proposals/permissions_spec.rb:
Why not using the traits in this spec? Should we add at least a comment to explain the reason?

ahukkanen replied:
I guess the reason was to ensure we have the user created with the same attributes in all specs.

Anyways, changed it as requested at: 3fcfb38b81be9f15f77d8723e463b4e1e84f6708

andreslucena commented on decidim-conferences/app/presenters/decidim/conference_speaker_presenter.rb:
Can you add specs for documenting this change 🙏🏽 ? If it is and I miss it please point out the file/context so I can check it out.

ahukkanen replied:
Improved test coverage at: e9502b7eb4cf8f6bd66b765f50e307fd1f324576

andreslucena commented on decidim-core/app/controllers/decidim/profiles_controller.rb:
Can you add specs for documenting this change 🙏🏽 ? If it is and I miss it please point out the file/context so I can check it out.

ahukkanen replied:
https://github.com/mainio/decidim/blob/b991536be591155bd5cb67ac513aa1d478bd4bc6/decidim-core/spec/controllers/decidim/profiles_controller_spec.rb#L57-L63

andreslucena reviewed (commented):
I tried the happy path (creating an account, not confirming, seeing that the profile isn't published, confirming the account, seeing that the profile is published). 

Apart from what I found in the first round, can you add a Release Note for this? It should serve as a heads-up for implementers/developers in case. Thanks

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Remove mention about user groups in comment

ahukkanen committed:
Rename `PersonalUserPresenter` to `SessionUserPresenter`

ahukkanen committed:
Use traits instead of updating the existing record

ahukkanen committed:
Improve test coverage for the conference speaker presenter

ahukkanen committed:
Fix the spec name for `session_user_presenter_spec`

ahukkanen committed:
Add release note

ahukkanen commented:
> can you add a Release Note for this?

Release note added at af748001c256ca2947273c3085073cfac40666e6.

andreslucena commented on decidim-core/spec/types/organization_type_spec.rb:
```diff
@@ -24,13 +24,14 @@ module Core
         let!(:unconfirmed_users) { create_list(:user, 2, organization: model) }
 
         it "shows all the stats for this organization" do
           # As we have 4 confirmed users and the admin (assuming they are counted as confirmed), we expect 5.
```
update the comment please

ahukkanen replied:
6d9d9a6230b76310140b0a4779c185e0d2610d85

andreslucena commented on decidim-core/spec/shared/profile_visibility_examples.rb:
Can you rename this file to `decidim-core/spec/shared/profile_visibility_examples.rb`? When the shared example title and the filename doesn't match it makes much difficult to find the files (you need to do a grep  + find file instead of just a find file)

ahukkanen replied:
0ea2293a579495f2e5ec1d668268305b29a8ebb2

andreslucena reviewed (dismissed):
Two nitpicks and this should be ready to the merge IMO

andreslucena commented:
As this is a big change, @alecslupu can you check this out too? If you don't have availability in the next few days let me know, I don't want to delay this even more (as it will get more conflicts, and it is already an old PR that I want to merge). Thanks

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Remove unnecessary spec comment

Removes the comment as the `current_user` in this spec is set to
`nil`, and therefore the amount of participants is as defined by
the spec context.

ahukkanen committed:
Rename spec shared example to represent the name of the example

ahukkanen committed:
Fix N+1 in the report model validation

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

alecslupu reviewed (commented):
I am looking at this PR, and i am not 100% that we can merge. 

I have a few doubts on the following: 
- We have around 800, rows that responds to `grep -RiHn 'create(:user' ./ | grep -v ':confirmed'` , which means we have around 800 objects that may not properly test the feature added
- The user factory has some children (process_admin, process_collaborator, process_moderator, process_evaluator + assembly related) , which do not test if the user is confirmed or not ... 
- Any elevated role (admin, moderator, evaluator etc) should have permissions if the account is not confirmed ?

andreslucena commented:
@ahukkanen can you check Alex's comment please? Thanks

ahukkanen committed:
Add core profile visibility regression coverage

ahukkanen committed:
Add profile visibility regression coverage

ahukkanen committed:
Add more specs for user presenter

ahukkanen committed:
Add test for the user search indexing visibility

Test that only the confirmed accounts are indexed.

ahukkanen committed:
Merge branch 'develop' into feature/profile-publicity-scopes

ahukkanen committed:
Make sure the dummy author has accepted tos to be visible

ahukkanen committed:
Fix adding a follow to invisible author

ahukkanen committed:
Add more tests for fetching invisible accounts through the API

ahukkanen committed:
Add test for searching users (confirmed and unconfirmed)

ahukkanen committed:
Test that the coauthorship adds the follow to the user author

ahukkanen commented:
Most of the spec files listed by the grep command are unrelated and profile visibility is not a concern in them. They test e.g. commands, permissions, jobs, mailers, imports, forms, serializers, admin side of things, etc. There is not much sense extending the tests in these cases as profile visibility (confirmed / unconfirmed) is not a concern in those.

I have extended the test cases for the author cell, user presenter, user model search indexing and the API related to fetching user / author records where possibly more test cases make sense. I also fixed one edge case bug if the a coauthor is unconfirmed, it did not add the follow correctly (would not happen in a normal situation, as you'd be confirmed when added as a coauthor).

Rather than doing a global grep without any context or reasoning, it would be more useful to point out what else do you want to be tested for this? I believe most cases where profile visibility matters are already covered, such as the User model (now including its search indexing), profiles controller, user activities controller, user presenter, author cell and the API.

Not saying there couldn't be bugs still, but that's the situation with almost any feature.

Which tests do you think are missing and why? Please note that this PR deals with user visibility when the account is unconfirmed. It is not a concern regarding user visibility if e.g. unconfirmed account is able to run a command or act as the current user of a form (as in many those examples).

alecslupu reviewed (approved):
👍 I cannot say anything more than i already said. I understand that this PR can introduce bugs, but i cannot identify now ... Let's merge it, and fix whatever happens at a later stage. 

@ahukkanen Thank you for your  effort.
